pkg/hubble/exporter/dynamic_exporter.go

```diff
@@ -0,0 +1,192 @@
+	"github.com/sirupsen/logrus"
+
+	v1 "github.com/cilium/cilium/pkg/hubble/api/v1"
+	"github.com/cilium/cilium/pkg/hubble/exporter/exporteroption"
+	"github.com/cilium/cilium/pkg/hubble/metrics"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+const gaugeMetricReportInterval = 5 * time.Second
+
+type dynamicExporter struct {
+	logger           logrus.FieldLogger
+	watcher          *configWatcher
+	managedExporters map[string]*managedExporter
+	maxFileSizeMB    int
+	maxBackups       int
+	mutex            lock.RWMutex
+	metricsTicker    *time.Ticker
+	metricsStop      chan bool
+}
+
+// OnDecodedEvent distributes events across all managed exporters.
+func (d *dynamicExporter) OnDecodedEvent(ctx context.Context, event *v1.Event) (bool, error) {
+	select {
+	case <-ctx.Done():
+		return false, d.Stop()
+	default:
+	}
+
+	d.mutex.RLock()
+	defer d.mutex.RUnlock()
+	for _, me := range d.managedExporters {
+		if me.config.End.After(time.Now()) {
+			me.exporter.OnDecodedEvent(ctx, event)
+		}
+	}
+	return false, nil
+}
+
+// Stop stops configuration watcher, metrics propagation and all managed flow
+// log exporters.
+func (d *dynamicExporter) Stop() error {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	d.watcher.Stop()
+	d.metricsTicker.Stop()
+	d.metricsStop <- true
+
+	var accErrs []error
+	for _, me := range d.managedExporters {
+		accErrs = append(accErrs, me.exporter.Stop())
+	}
+
+	return errors.Join(accErrs...)
+}
+
+// NewDynamicExporter creates instance of dynamic hubble flow exporter.
+func NewDynamicExporter(configFilePath string, maxFileSizeMB, maxBackups int) *dynamicExporter {
+	dynamicExporter := &dynamicExporter{
+		logger:           logrus.New(),
+		managedExporters: make(map[string]*managedExporter),
+		maxFileSizeMB:    maxFileSizeMB,
+		maxBackups:       maxBackups,
+	}
+
+	watcher := NewConfigWatcher(configFilePath, dynamicExporter.onConfigReload)
+	dynamicExporter.watcher = watcher
+	dynamicExporter.startMetricReporting()
+	return dynamicExporter
+}
+
+func (d *dynamicExporter) startMetricReporting() {
+	d.metricsTicker = time.NewTicker(gaugeMetricReportInterval)
+	d.metricsStop = make(chan bool)
+
+	go func() {
+		for {
+			select {
+			case <-d.metricsStop:
+				return
+			case <-d.metricsTicker.C:
+				d.reportGaugeMetrics()
+			}
+		}
+	}()
+}
+
+func (d *dynamicExporter) reportGaugeMetrics() {
+	d.mutex.RLock()
+	defer d.mutex.RUnlock()
+
+	var activeExporters, inactiveExporters float64
+
+	for name, me := range d.managedExporters {
+		var value float64
+		if me.config.End.After(time.Now()) {
+			value = 1
+			activeExporters++
+		} else {
+			value = 0
+			inactiveExporters++
+		}
+		metrics.IndividualExportersStatus.WithLabelValues(name).Set(value)
+	}
+
+	metrics.Exporters.WithLabelValues("active").Set(activeExporters)
+	metrics.Exporters.WithLabelValues("inactive").Set(inactiveExporters)
+}
+
+func (d *dynamicExporter) onConfigReload(ctx context.Context, config dynamicExportersConfig) {
```

Comment:
There doesn't seem to be any validation that the `Name` is unique. What happens if someone specifies the same `name` twice?

Reply:
Also, what about the file path? What if you specify the same file path twice?

Reply:
added validation